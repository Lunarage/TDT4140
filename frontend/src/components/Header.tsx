import { useEffect, useState } from "react";
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from "react-router-dom";
import styled from "styled-components";
import { logoColor, pageName, redHexColor } from "../consts";
import { getAdminStatistics } from '../store/actionCreators';
import { State } from '../store/types';
import { ExpandWrapper } from './ActivityDashboard';
import Button from './Button';
import Loading from './Loading';
import NewActivity from './NewActivity';

const Wrapper = styled.div`
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const UserButton = styled.div`
  display: flex;
  margin: 10px;

  &:hover {
    cursor: pointer;
  }
`;

const Tab = styled.div`
  display: flex;
  font-size: 20px;
  margin: 10px 15px;
  color: ${redHexColor};
  font-weight: bold;

  &:hover {
    cursor: pointer;
  }
`;

const TabsWrapper = styled.div`
  display: flex;
  align-items: baseline;
`;

export const Logo = styled.div`
  margin: 10px 20px;
  font-size: 30px;
  font-weight: bold;
  color: ${logoColor};

  &:hover {
    cursor: pointer;
  }
`;

const NewActivityResponse = styled.div`
  background-color: ${redHexColor};
  border: solid;
  border-radius: 10px;
  border-color: white;
  width: auto;
  min-width: 150px;
  height: auto;
  min-height: 100px;
  box-shadow: 1px 1px 20px 4px rgba(0, 0, 0, 0.45);
  color: white;
  font-size: 30px;
  position: absolute;
  padding: 1em;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
  display: flex;
  justify-content: center;
  align-items: center;
`;

export enum Type {
  aktiviteter = "Aktiviteter",
  arrangementer = "Arrangementer"
}

// General page header used on every page
const Header = () => {
  const history = useHistory();
  const dispatch = useDispatch();

  const [showCreateNew, setShowCreateNew] = useState<boolean>(false);
  const [showSuccess, setShowSuccess] = useState<boolean>(false);

  const {
    stats,
    isLoading: statsIsLoading,
    errorMessage: statsError,
  } = useSelector((state: State) => state.statsReducer);

  useEffect(() => {
    if (!stats && !statsIsLoading && !statsError) {
      dispatch(getAdminStatistics())
    }
  }, [stats])

  const {
    event,
    isLoading: eventLoading,
    errorMessage: eventError,
  } = useSelector((state: State) => state.postEventReducer);

  const setUrl = (tab: string) => {
    history.push(tab);
  };

  const handleTypeClick = (tab: string, type: string) => {
    history.push(tab + "?type=" + type); // adds aktiviteter or arrangementer to url
  };

  const handleNewActivityExit = (submit: boolean) => {
    setShowCreateNew(false);
    if (submit) {
      setShowSuccess(true)
      setTimeout(() => setShowSuccess(false), 4000)
    }
  }

  const handleLogOut = () => {
    localStorage.clear()
    setUrl("/")
    window.location.reload()
    return false
  }

  if (eventLoading) return <Loading />

  return (
    <>
      {/* displays successs or failure of creating new activity */}
      {showCreateNew && <ExpandWrapper > <NewActivity onExitFunc={handleNewActivityExit} /></ExpandWrapper>}
      {!showCreateNew && showSuccess && (eventError ?
        <NewActivityResponse>Klarte ikke Ã¥ poste aktiviteten.</NewActivityResponse> :
        event && <NewActivityResponse>{event?.title} er postet!</NewActivityResponse>)}
      <Wrapper>
        <Logo onClick={() => setUrl("/")}>{pageName}</Logo>
        <TabsWrapper>
          <Tab onClick={() => handleTypeClick("/browse", "aktiviteter")}>
            {Type.aktiviteter}
          </Tab>
          <Tab
            onClick={() => handleTypeClick("/browse", "arrangementer")}
          >
            {Type.arrangementer}
          </Tab>
          {localStorage.getItem("token") && <Button
            text="Lag ny"
            onClickFunc={() => setShowCreateNew(true)}
          />}
        </TabsWrapper>
        {localStorage.getItem("token") ? (
          <TabsWrapper>
            {stats &&
              <UserButton onClick={() => setUrl("/admin")}>
                Statistikk
          </UserButton>}
            <UserButton onClick={() => setUrl("/mypage")}>
              Min side
          </UserButton>
            <UserButton onClick={handleLogOut}>
              Logg ut
            </UserButton>
          </TabsWrapper>
        ) : (
          <UserButton onClick={() => setUrl("/login")}>
            Logg inn
          </UserButton>
        )}
      </Wrapper>
    </>
  );
};

export default Header;
