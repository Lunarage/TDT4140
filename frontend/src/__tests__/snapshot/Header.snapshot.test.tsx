import { mount } from "enzyme";
import Enzyme from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJSON from "enzyme-to-json";
import Header from '../../components/Header';
import { store } from '../..';
import { Provider } from 'react-redux';

Enzyme.configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  ...jest.requireActual('react-router-dom') as any,
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

const HeaderWrapper = () => (
  <Provider store={store}>
    <Header />
  </Provider>
);

describe("Header", () => {
  it("renders Header", () => {
    mount(<HeaderWrapper />);
  });
  it("Should match Header snapshot", () => {
    const comp = mount(<HeaderWrapper />);
    expect(toJSON(comp)).toMatchSnapshot();
  });
});

