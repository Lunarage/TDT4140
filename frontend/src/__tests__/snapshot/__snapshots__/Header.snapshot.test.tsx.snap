// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Should match Header snapshot 1`] = `
<HeaderWrapper>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Header>
      <styled.div>
        <div
          className="sc-iktFzd uGybn"
        >
          <styled.div
            onClick={[Function]}
          >
            <div
              className="sc-cBNfnY bDxDMJ"
              onClick={[Function]}
            >
              Gj√∏r No'
            </div>
          </styled.div>
          <styled.div>
            <div
              className="sc-gWHgXt bcglBl"
            >
              <styled.div
                onClick={[Function]}
              >
                <div
                  className="sc-hiSbYr hZBClG"
                  onClick={[Function]}
                >
                  Aktiviteter
                </div>
              </styled.div>
              <styled.div
                onClick={[Function]}
              >
                <div
                  className="sc-hiSbYr hZBClG"
                  onClick={[Function]}
                >
                  Arrangementer
                </div>
              </styled.div>
            </div>
          </styled.div>
          <styled.div
            onClick={[Function]}
          >
            <div
              className="sc-jJEJSO cgHETT"
              onClick={[Function]}
            >
              Logg inn
            </div>
          </styled.div>
        </div>
      </styled.div>
    </Header>
  </Provider>
</HeaderWrapper>
`;
